transmart-core-db/grails-app/services/org/transmartproject/db/userqueries/UserQueryDiffService.groovy

```diff
@@ -0,0 +1,242 @@
+package org.transmartproject.db.userqueries
+
+import grails.converters.JSON
+import grails.transaction.Transactional
+import org.grails.web.converters.exceptions.ConverterException
+import org.grails.web.json.JSONException
+import org.hibernate.Criteria
+import org.hibernate.SessionFactory
+import org.hibernate.criterion.DetachedCriteria
+import org.hibernate.criterion.Order
+import org.hibernate.criterion.Projections
+import org.hibernate.criterion.Property
+import org.hibernate.criterion.Restrictions
+import org.hibernate.sql.JoinType
+import org.springframework.beans.factory.annotation.Autowired
+import org.transmartproject.core.exceptions.AccessDeniedException
+import org.transmartproject.core.exceptions.InvalidArgumentsException
+import org.transmartproject.core.exceptions.NoSuchResourceException
+import org.transmartproject.core.querytool.QueryResult
+import org.transmartproject.core.userquery.UserQuery
+import org.transmartproject.core.userquery.UserQueryDiffEntry
+import org.transmartproject.core.userquery.UserQueryDiffResource
+import org.transmartproject.core.users.User
+import org.transmartproject.db.clinical.MultidimensionalDataResourceService
+import org.transmartproject.db.multidimquery.query.Constraint
+import org.transmartproject.db.multidimquery.query.ConstraintFactory
+import org.transmartproject.db.querytool.Query
+import org.transmartproject.db.querytool.QueryDiff
+import org.transmartproject.db.querytool.QueryDiffEntry
+import org.transmartproject.db.user.User as DbUser
+import org.transmartproject.core.users.UsersResource
+import org.transmartproject.db.accesscontrol.AccessControlChecks
+
+@Transactional
+class UserQueryDiffService implements UserQueryDiffResource {
+
+    @Autowired
+    UsersResource usersResource
+
+    @Autowired
+    UserQueryService userQueryService
+
+    @Autowired
+    MultidimensionalDataResourceService multiDimService
+
+    @Autowired
+    AccessControlChecks accessControlChecks
+
+    SessionFactory sessionFactory
+
+    @Override
+    void scan(User currentUser) {
+        DbUser user = (DbUser) usersResource.getUserFromUsername(currentUser.username)
+        if (!user.admin) {
+            throw new AccessDeniedException('Only allowed for administrators.')
+        }
+        // get list of all not deleted queries per user
+        List<UserQuery> userQueries = userQueryService.list()
+        if (!userQueries) {
+            log.info "No queries were found."
+            return
+        }
+
+        for (query in userQueries) {
+            DbUser queryUser = (DbUser) usersResource.getUserFromUsername(query.username)
+            def oldSet = getOldSet(query, queryUser)
+            if (!oldSet) {
+                log.info "Set for query: '$query.id' was not found."
+                return
```

Comment:
Why return? I would say this initial diff contains all entries.

Reply:
I'm not sure, all new entries would be send in an email with daily/weekly updates and what I understood is that this should contain information only about sets that where changed because of the data change (after data loading). I'll ask for that to make sure.